core/src/main/java/org/apache/calcite/rel/rules/SingleValuesOptimizationRules.java

```diff
@@ -0,0 +1,374 @@
+      this.bitSet = bitSet;
+      this.fieldValues = values;
+      this.offset = offset;
+    }
+    @Override public RexNode visitInputRef(RexInputRef inputRef) {
+      if (bitSet.get(inputRef.getIndex())) {
+        return this.fieldValues.get(inputRef.getIndex() + offset);
+      }
+      return super.visitInputRef(inputRef);
+    }
+
+    public RexNode go(RexNode condition) {
+      return condition.accept(this);
+    }
+  }
+
+  /**
+   * Abstract prune single value rule that implements SubstitutionRule interface.
+   */
+  protected abstract static class PruneSingleValueRule
+      extends RelRule<PruneSingleValueRule.Config>
+      implements SubstitutionRule {
+    protected PruneSingleValueRule(PruneSingleValueRule.Config config) {
+      super(config);
+    }
+
+    protected BiFunction<RexNode, List<RexNode>, List<RexNode>>
+        getRexTransformer(RexBuilder rexBuilder,
+        JoinRelType joinRelType) {
+      switch (joinRelType) {
+      case LEFT:
+      case RIGHT:
+        return (condition, rexLiterals) -> rexLiterals.stream().map(lit ->
+            rexBuilder.makeCall(SqlStdOperatorTable.CASE, condition,
+                lit, rexBuilder.makeNullLiteral(lit.getType()))).collect(Collectors.toList());
+      default:
+        return (condition, rexLiterals) -> new ArrayList<>(rexLiterals);
+      }
+    }
+
+    /**
+     * onMatch function contains common optimization logic for all the
+     * SingleValueOptimization rules. It simplifies the rel node tree by
+     * removing a Join node and creating a required filter as applicable.
+     * In case of the LEFT/RIGHT joins, a case expression which produce NULL
+     * values for non-matching rows will be created as part of a Project node.
+     *
+     * @param call A RelOptRuleCall object
+     * @param values A constant relation node which produces a single row.
+     * @param project A project node which has dynamic constants (can be null).
+     * @param join A join node which will get removed.
+     * @param other A node on the other side of the join (apart from Values)
+     * @param isLeft Whether a Values node is on the Left side or the right side of the Join.
+     */
+    protected void onMatch(RelOptRuleCall call, Values values,
+        @Nullable Project project, Join join,
+        RelNode other, boolean isLeft) {
+      Predicate<Join> predicate = eligibilityPredicate(isLeft);
+      List<RexNode> rexNodes;
+      if (project != null) {
+        ImmutableBitSet bitSet = ImmutableBitSet.range(0, values.getRowType().getFieldCount());
+        RexShuttle shuttle =
+            new ReplaceExprWithConstValue(bitSet,
+                    new ArrayList<>(values.getTuples().get(0)),
+                0);
+
+        rexNodes = project.getProjects().stream()
+            .map(shuttle::apply)
+            .collect(Collectors.toList());
+      } else {
+        rexNodes = new ArrayList<>(values.tuples.get(0));
```

Comment:
Maybe you can factor out this as a helper function.
You previously used getTuples(), here you just used tuples.
This could be called getSingle(Values) and could also assert that values has a single tuple.

Reply:
I changed the code to use tuples in both the cases. I haven't created a helper function here because it might be an overkill.